### ISSUE TITLE
Optim.Adam 'step' default setting bug.

### ISSUE THREAD

**Issue by 1996wu** [pytorch/pytorch#110940]
### 🐛 Describe the bug

Optim.Adam 'step' default setting bug.

```python
import torch
import torch.nn as nn
import torch.optim as optim

device = "cuda"
dtype = torch.float64
torch.set_default_dtype(torch.float64)

X = torch.randn(10, 1, device=device, dtype=dtype)
y = 2 * X + 1 + 0.1 * torch.randn(10, 1, device=device, dtype=dtype)  

class LinearRegression(nn.Module):
    def __init__(self, device: str = None):
        super(LinearRegression, self).__init__()
        self.linear = nn.Linear(1, 1, device=device)
    def forward(self, x):
        return self.linear(x)

model = LinearRegression(device=device)
criterion = nn.MSELoss()
optimizer = optim.Adam(model.parameters(), lr=0.01)

for epoch in range(10):
    outputs = model(X)
    loss = criterion(outputs, y)
    optimizer.zero_grad()
    loss.backward()
    optimizer.step()
```

```
Traceback (most recent call last):
  File "/home/zbwu/Desktop/example/test-1.py", line 32, in <module>
    optimizer.step()
  File "/home/zbwu/soft/anaconda3/lib/python3.10/site-packages/torch/optim/optimizer.py", line 373, in wrapper
    out = func(*args, **kwargs)
  File "/home/zbwu/soft/anaconda3/lib/python3.10/site-packages/torch/optim/optimizer.py", line 76, in _use_grad
    ret = func(self, *args, **kwargs)
  File "/home/zbwu/soft/anaconda3/lib/python3.10/site-packages/torch/optim/adam.py", line 165, in step
    adam(
  File "/home/zbwu/soft/anaconda3/lib/python3.10/site-packages/torch/optim/adam.py", line 313, in adam
    func(params,
  File "/home/zbwu/soft/anaconda3/lib/python3.10/site-packages/torch/optim/adam.py", line 476, in _multi_tensor_adam
    grouped_tensors = Optimizer._group_tensors_by_device_and_dtype(
  File "/home/zbwu/soft/anaconda3/lib/python3.10/site-packages/torch/optim/optimizer.py", line 397, in _group_tensors_by_device_and_dtype
    return _group_tensors_by_device_and_dtype(tensorlistlist, with_indices)
  File "/home/zbwu/soft/anaconda3/lib/python3.10/site-packages/torch/utils/_contextlib.py", line 115, in decorate_context
    return func(*args, **kwargs)
  File "/home/zbwu/soft/anaconda3/lib/python3.10/site-packages/torch/utils/_foreach_utils.py", line 47, in _group_tensors_by_device_and_dtype
    torch._C._group_tensors_by_device_and_dtype(tensorlistlist, with_indices).items()
RuntimeError: Tensors of the same index must be on the same device and the same dtype except `step` tensors that can be CPU and float32 notwithstanding
```

The optimizer Adam 'step' dtype is set to float64, which leads to the above error. 
 If default 'step' is changed to **torch.tensor(0., dtype=torch.float)** or cancel  **torch.set_default_dtype(torch.float64)** to avoid the above situation.

### Versions

Collecting environment information...
PyTorch version: 2.1.0
Is debug build: False
CUDA used to build PyTorch: 11.8
ROCM used to build PyTorch: N/A

OS: Arch Linux (x86_64)
GCC version: (Arch Linux 10.5.0-1) 10.5.0
Clang version: 16.0.6
CMake version: version 3.27.7
Libc version: glibc-2.38

Python version: 3.10.13 (main, Sep 11 2023, 13:44:35) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-6.1.55-1-lts-x86_64-with-glibc2.38

cc @vincentqb @jbschlosser @albanD @janeyx99 @crcrpar

**Comment by YOUSEFNANIS** [comment]
I tried the same code on python 3.10, windows 11, intel CPU, and it works fine
the problem might be related to cuda

**Comment by janeyx99** [comment]
@crcrpar I’m inclined to think we should either be supporting step as float32 or float64 OR we should hardcode dtype to float32. What are your thoughts?

**Comment by jon-chuang** [comment]
I think the contract right now is that step is always float32, hence I think that the current impl is actually implicitly relying on float32 default dtype, but given the default can be changed, we now need to be explicit.

**Comment by crcrpar** [comment]
theoretically there can be `step` that overflows float32 so allowing for float64 `step` could make sense as well but I second @jon-chuang's approach (at least for now), partly because `fused` implementations have hardcoded float32: https://github.com/pytorch/pytorch/blob/95ff51d8ede577c9ac1fbcbbe4a765156e028f91/torch/optim/adam.py#L105-L109

**Comment by janeyx99** [comment]
Reopening due to Richard's comment. @richardk53 I'd like to understand the use case for a float64 step a bit more. What are some applications where float64 is necessary?

**Comment by viktor-ktorvi** [comment]
I had the same issue and my use case is ML for approximating optimization problems. 

In my current setting I try to learn the outputs of a solver (the solver uses float64). Taking the solvers solution and casting it to float32 yields unacceptably large errors in the equality constraints. The equality constraint functions are complicated and functions of many variables so the little errors between 32 and 64 add up to produce larger ones in the function output.

For that reason the model needs to take in float64 inputs and to output float64.

In short, the use case is in ML for science or optimization.

**Comment by janeyx99** [comment]
In that case, I would propose the following actionable items:

1. Allow step to be float64 if the default type is float64 for the default + foreach implementations
2. Allow fused adam to support float64 step--this would be less trivial, cc @crcrpar

**Comment by janeyx99** [comment]
Just opened a PR for step 1. This should allow for float64 step (and other scalars) for the non fused implementations.

**Comment by janeyx99** [comment]
@viktor-ktorvi @richardk53 I've just landed a PR to enable float64 on default optimizers. Let me know if you're able to give it a try through our nightlies. (One can install the nightly from the Install instructions on https://pytorch.org/.)

**Comment by viktor-ktorvi** [comment]
I'll give it a try soon!

**Comment by viktor-ktorvi** [comment]
Hi,

I think it looks good. The model input, output and weights are all float64 and the step works, which is what I would expect should happen. I haven't been able to integrate this into my use case because my dependencies support up to torch 2.1.*, but I think that'll all be good. Thanks!

**Comment by janeyx99** [comment]
Thanks for verifying! I’ll close this issue and have opened https://github.com/pytorch/pytorch/issues/116592 to track the fused case.

**Comment by Oscarjia** [comment]
may i ask how can i quick resolve the bug? i have used  
`nvcr.io/nvidia/pytorch:23.11-py3`

and the installed pytorch version is 
`torch-2.2.0a0%2B6a974be-cp310-cp310-linux_x86_64.whl`
should i reinstall pytorch to specific version?

**Comment by janeyx99** [comment]
Yes, 6a974be was committed in Oct 2023 and the fix for this went in afterwards at the end of December. Could you try again with the latest torch with `pip install`? The latest nightly versions can also be installed by following the toggle thingie in the front pytorch page: https://pytorch.org/

**Comment by Oscarjia** [comment]
Hi, @janeyx99, I really appreciate your prompt response! However, I'm currently unable to test the latest nightly version because my currently work relies on `torch-2.2.0a0+6a974be-cp310-cp310-linux_x86_64.whl`, which is also a dependency for the `flash-attn` package. 

So I was wondering, if I manually update all the files under `torch/optim` as per the modifications in commit [62de29](https://github.com/pytorch/pytorch/commit/62de29d06f1ddcc7c4b11757adc75f1459ef6991), would that resolve the issue?

**Comment by janeyx99** [comment]
yes you can try patching it if that is the easier thing to do!

**Comment by Oscarjia** [comment]
@janeyx99  Great! I will try and let you know the results.
